pymatgen/core/surface.py

```diff
@@ -1217,105 +1217,122 @@ def get_z_ranges(
-                            if nn[0].species_string == pair[idx - 1]:
+                    if site.species_string == ele:
+                        for nn in self.oriented_unit_cell.get_neighbors(site, bond_dist):
+                            if nn[0].species_string == species_pair[idx - 1]:
                                 poly_coord += 1
+
                     cn_list.append(poly_coord)
-                cn_dict[el] = cn_list
+                cn_dict[ele] = cn_list
 
-            # We make the element with the higher coordination our reference
-            if max(cn_dict[pair[0]]) > max(cn_dict[pair[1]]):
-                element1, element2 = pair
+            # Make the element with higher coordination the reference
+            if max(cn_dict[species_pair[0]]) > max(cn_dict[species_pair[1]]):
+                element1, element2 = species_pair
             else:
-                element2, element1 = pair
+                element2, element1 = species_pair
 
             for idx, site in enumerate(slab):
-                # Determine the coordination of our reference
+                # Determine the coordination of the reference
                 if site.species_string == element1:
                     poly_coord = 0
-                    for neighbor in slab.get_neighbors(site, bond_len):
+                    for neighbor in slab.get_neighbors(site, bond_dist):
                         poly_coord += 1 if neighbor.species_string == element2 else 0
 
-                    # suppose we find an undercoordinated reference atom
+                    # Suppose we find an undercoordinated reference atom
                     if poly_coord not in cn_dict[element1]:
-                        # We get the reference atom of the broken bonds
+                        # Get the reference atom of the broken bonds
                         # (undercoordinated), move it to the other surface
                         slab = self.move_to_other_side(slab, [idx])
 
-                        # find its NNs with the corresponding
+                        # Find its NNs with the corresponding
                         # species it should be coordinated with
-                        neighbors = slab.get_neighbors(slab[idx], bond_len, include_index=True)
+                        neighbors = slab.get_neighbors(slab[idx], bond_dist, include_index=True)
                         to_move = [nn[2] for nn in neighbors if nn[0].species_string == element2]
                         to_move.append(idx)
-                        # and then move those NNs along with the central
+                        # Move those NNs along with the center
                         # atom back to the other side of the slab again
                         slab = self.move_to_other_side(slab, to_move)
 
         return slab
 
-    def move_to_other_side(self, init_slab: Slab, index_of_sites: list[int]) -> Slab:
-        """Move a set of sites to the opposite surface of the slab.
+    def move_to_other_side(
```

Comment:
There are multiple issues with this `move_to_other_side` method, ranked by severity from most to least: 

1. Wrong sign for the moving vector: 
`slab.translate_sites(top_site_index, vector=[0, 0, slab_ratio])`
where the `slab_ratio` is the slab/total_cell ratio (positive). This means the sites on the top is moved further upwards, instead of the intended downwards (the opposite side).

2. Confusing and imprecise calculation of moving vector `slab_ratio` as:
```python
  nlayers_slab: int = math.ceil(self.min_slab_size / height)
  nlayers_vac: int = math.ceil(self.min_vac_size / height)
  nlayers: int = nlayers_slab + nlayers_vac
  
  slab_ratio: float = nlayers_slab / nlayers
```
As of (1), currently the sites are moved by a vector `slab_ratio`, but using the slab/total_cell **layer** ratio is very confusing and imprecise(two roundings are included) in this use case. I would prefer just calculate the vector by the slab/cell **height ratio** instead of the **number of layers ratio**.

3. The use case of this method should be limited to **surface sites** instead of **any site** inside the Slab (need clarify in docstring). Reason being this method actually move the sites by a length equal to the thickness of the Slab, if people use it to move some site closer to the center, the site would then end up deep in the vacuum layer.

Reply:
thanks for taking a deep dive on this module. such doc string clarifications are very valuable!

Reply:
re (1), could you comment on why the tests did not catch this? could you add a new test first that demonstrates the bug?

https://github.com/materialsproject/pymatgen/blob/2a3608f31d44b7ed8c23f64b9c9bbe07f518a5ae/tests/core/test_surface.py#L561-L579

re (2): 

> I would prefer just calculate the vector by the slab/cell height ratio instead of the number of layers ratio.

would we lose backwards compat? if not, feel free to simplify/clarify the implementation

Reply:
> re (1), could you comment on why the tests did not catch this? could you add a new test first that demonstrates the bug?

That's a good point. I wasn't looking into the test (but I should). I would have a look into the test and update this comment. But I just tried to reverse the moving vector and the unit test still passed, so I assume the test didn't capture the issue.

> re (2):
> 
> > I would prefer just calculate the vector by the slab/cell height ratio instead of the number of layers ratio.
> 
> would we lose backwards compat? if not, feel free to simplify/clarify the implementation

I believe it would not. It's an internal "moving distance" calculation method, and I believe replacing it with another method would not bring up any compat issue (as it only affect how much the site is move when `move_to_other_side`, and current implement use the ratio of slab_layers versus total_cell_layers to determine the proportion of Slab inside the whole Cell).

Reply:
I decide maybe it's wise to separate such function tweaks into a separate PR. Current PR has too many changes already and it would be hard to notice any change/issue.